### ISSUE TITLE
multi_logloss differs between native and custom objective function (identical to native objective function)

### ISSUE THREAD

**Issue by kapytaine** [microsoft/LightGBM#4211]
## Description
I created two mutli class classifier models, one with the native objective function and one with a custom objective function. Before implementing my own custom objective function I tried to code the native objective function in order to validate my way of coding. I managed to get identical results. The trees are the same as well as the raw results. However the displayed multi_logloss during training differs from one model to the other.

## Reproducible example
```
# Packages
import numpy as np
import pandas as pd

from lightgbm import LGBMClassifier

# Data
N_train = 1000
N_test = 100
X_train = pd.DataFrame(
    {
        "x1": np.random.normal(loc=0, scale=10.0, size=N_train),
        "x2": np.random.normal(loc=0, scale=10.0, size=N_train),
    }
)
y_train = np.random.choice(["0", "1", "2"], replace=True, size=N_train)
X_test = pd.DataFrame(
    {
        "x1": np.random.normal(loc=0, scale=10.0, size=N_test),
        "x2": np.random.normal(loc=0, scale=10.0, size=N_test),
    }
)
y_test = np.random.choice(["0", "1", "2"], replace=True, size=N_test)

# model 1 with native objective function
clf1 = LGBMClassifier(
    objective="multiclass",
    n_estimators=10,
    n_jobs=1,
    max_depth=3,
    reg_lambda=1,
)
clf1.fit(
    X_train,
    y_train,
    eval_set=[(X_test, y_test)],
    eval_metric=["multi_logloss", "multi_error"],
    init_score=[0] * len(X_train) * 3,
)
raw_scores_1 = clf1.predict(X_test, raw_score=True)

# model 2 with custom objective function
def custom_objective_function(y_true, y_pred):
    n_trains = len(y_true)
    n_classes = int(len(y_pred) / n_trains)

    y_pred_reshaped = np.reshape(y_pred, (n_trains, n_classes), order="F")

    y_true_reshaped = np.zeros_like(y_pred_reshaped)
    y_true_reshaped[range(y_pred_reshaped.shape[0]), y_true.astype(int)] = 1

    denominator = np.sum(np.exp(y_pred_reshaped), axis=1, keepdims=True)
    prob_pred = np.exp(y_pred_reshaped) / denominator

    grad = (prob_pred - y_true_reshaped).ravel(order="F")
    hess = 1.5 * (prob_pred * (1 - prob_pred)).ravel(order="F")

    return grad, hess

clf2 = LGBMClassifier(
    objective=custom_objective_function,
    n_estimators=10,
    max_depth=3,
    n_jobs=1,
    reg_lambda=1,
)
clf2.fit(
    X_train,
    y_train,
    eval_set=[(X_test, y_test)],
    eval_metric=["multi_logloss", "multi_error"],
    init_score=[0] * len(X_train) * 3,
)
raw_scores_2 = clf2.predict(X_test, raw_score=True)
np.array_equal(raw_scores_1, raw_scores_2)
```

## Environment info

LightGBM version or commit hash: 3.2.1

Command(s) you used to install LightGBM

```shell
pip install lightgbm
```

## Additional Comments
Besides my initial question regarding the multi_logloss I have 2 interrogations:
- I managed to get identical results with the factor 1.5 for the hessian. I saw that there is a 2 coefficient in XGBoost. What is the motivation of using a factor for the hessian?
- If I don't set init_score to zeros I don't get the same results. Does using a custom objective function change the init scores?

**Comment by shiyu1994** [comment]
This is because LightGBM boosts from an initial `average` score according to the objective function. For example, with L2 regression, LightGBM would boost from the label mean of all training data. Setting `boost_from_average=False` in the parameters should produce a consistent result.

Please see previous similar issues like #4077.

**Comment by kapytaine** [comment]
@shiyu1994 thank you for your response.

Regarding my first question, do you know if there is a reason why the displayed multi_logloss values during training are different between the two trainings. For instance, after setting the seed to 0 before creating the datasets, I get the following results. The displayed multi_logloss in the second model (with custom objective function) is not what I expected. I had to code a custom eval_metric, custom_multilogloss.

> [1]	valid_0's multi_logloss: 1.09903	valid_0's multi_error: 0.75
> [2]	valid_0's multi_logloss: 1.09891	valid_0's multi_error: 0.71
> [3]	valid_0's multi_logloss: 1.09816	valid_0's multi_error: 0.71
> [4]	valid_0's multi_logloss: 1.09912	valid_0's multi_error: 0.71

> [1]	valid_0's multi_error: 0.75	valid_0's multi_logloss: 25.1089	valid_0's custom_multilogloss: 1.09903
> [2]	valid_0's multi_error: 0.71	valid_0's multi_logloss: 20.238	valid_0's custom_multilogloss: 1.09891
> [3]	valid_0's multi_error: 0.71	valid_0's multi_logloss: 20.9717	valid_0's custom_multilogloss: 1.09816
> [4]	valid_0's multi_error: 0.71	valid_0's multi_logloss: 23.2753	valid_0's custom_multilogloss: 1.09912

**Comment by shiyu1994** [comment]
For your first question, LightGBM uses the objective function to determine how to convert from raw scores to output. But with customized objective function (`objective ` in the following code snippet will be nullptr), no convert method can be specified. So the raw output will be directly fed to the metric function for evaluation. Though I think this can cause some unexpected results, as in your case. I think this should be marked as a bug.
https://github.com/microsoft/LightGBM/blob/1e95cb097757996cd82699071c82c205bf2b48a5/src/metric/multiclass_metric.hpp#L64-L116

For your first additional comment, the following code comment provides a explanation.
https://github.com/microsoft/LightGBM/blob/1e95cb097757996cd82699071c82c205bf2b48a5/src/objective/multiclass_objective.hpp#L26-L32

**Comment by kapytaine** [comment]
Thank you for your time and answers!
